### DISCUSSION TITLE
v3.0alpha1

### DISCUSSION THREAD

**Discussion by blackadderkate** [blackadderkate/weather-widget-2#161] [Announcements]
### A Beta version of the Weather Widget for KDE Plasma 6.

QT6 is MASSIVELY different to QT5. This required a substantial rewrite of this widget - in addition to the major issues below, there will be a lot of bugs to squash and polishing required...

**Major Bugs/Features/Issues to be sorted**
1) You can no longer access local FILE:/// urls using XMLHttpRequest. This breaks the MetNo Search. Two possible fixes:

1. host on a free web host - the downside is limited bandwidth allowances...
2. set QML_XHR_ALLOW_FILE_READ=1 - less secure and requires manual hacking by users

2)  Configuration Settings might not display or apply correctly.

3)  Locale-specific translations may be missing/incorrect.

4)  Some of the text doesn't appear to clip/elide correctly in the tables in the Configuration Pages.

5)  The Meteogram or Long-Range 6-hourly forecasts below might display using the wrong times, or display the wrong day/night icon, on certain combinations of system timezone and widget settings.

<hr /><em>This discussion was created from the release <a href='https://github.com/blackadderkate/weather-widget-2/releases/tagv3.0b1/'>v3.0b1</a>.</em>

**Comment by blackadderkate** [comment]
Please report minor bugs here, unless they're show-stoppers (like it not working at all)...

**Comment by blackadderkate** [comment]
I committed a few tweaks and bugfixes today.

Still TODO before Beta release:

- add Scrollboxes to the Config pages.

- update translations and test.

**Comment by blackadderkate** [comment]
Few more bugfixes.
Data wasn't being cached correctly for some reason.

Moving places up & down in the config did not get saved / applied.

**Comment by serhumanos** [comment]
how can I install this version to test it? thank you!

**Reply by blackadderkate** [reply]
First download the [latest test version for Plasma 6](https://github.com/blackadderkate/weather-widget-2/releases/download/v3.0a7/weather-widget-3.0a7.plasmoid) rom the releases section.
**Make sure you read the release notes for known bugs.**

Then follow the instructions [on this wiki page ](https://github.com/blackadderkate/weather-widget-2/wiki/How-to-install-the-Widget---manual-download) to install it.

**Comment by henricoetzee** [comment]
![image](https://github.com/blackadderkate/weather-widget-2/assets/97385924/9392ae10-d7dc-44ef-a2d5-8dfa72866b1a)

Temperatures are displayed in dark text with Breeze Dark Theme and Breeze light colors.

**Reply by blackadderkate** [reply]
I uploaded a fix for that yesterday - along with fixes for some of the Config pages, which also used the wrong colours. [download it here](https://github.com/blackadderkate/weather-widget-2/releases/download/v3.0b1b/weather-widget-3.0.0b1.plasmoid).

You might need to uninstall / delete the old version first...

**Reply by henricoetzee** [reply]
![image](https://github.com/blackadderkate/weather-widget-2/assets/97385924/d4c58ba4-55f6-49bc-9ab7-785a0fd44758)

Thanks.  I uninstalled the old one and installed the newer one, but the issue persists.

**Reply by blackadderkate** [reply]
Have you tried logging out and back in again, or restarting plasma (in konsole, **plasmashell - - replace &**)

**Reply by henricoetzee** [reply]
Rebooting did the trick. Thanks.

**Comment by dexter2000** [comment]
hello I'm seeing lots of 

`file:///home/dexter/.local/share/plasma/plasmoids/org.kde.weatherWidget-3/contents/ui/providers/OpenWeatherMap.qml:622:9: QML XmlListModelRole: Query error: "Attribute value not found"`

in the log. 
Fedora 40 
3.0.0b1.plasmoid OWM

**Comment by blackadderkate** [comment]
Hi,

If no rain is forecast during a specific time period, the Openweather data
feed omits the XML "precipitation" key altogether (rather than using Null
or Zero values), which the Qt XML parser doesn't like and flags as an error.

Due to changes in the way the XML engine works in Qt6, I can't access the
raw XML data any more - so there's no way to error trap this.

The only fix is to rewrite the Openweather provider to use Json instead of
XML - but that would be a major rewrite which would have delayed the
release of the widget even more, so I decided to release it with those
"errors".

For what it's worth, the XML feed has been deprecated by Openweather and I
have already added an issue to port it to Json once the Plasma 6 teething
troubles have been ironed out.

**Reply by dexter2000** [reply]
ok thanks for the update & all the work done so far. :-)

**Comment by martin-ms** [comment]
First of all, thank you for your work on the widget and for keeping the project alive.

I tried the installation your recommended way but failed in both cases. When I use the command line, after a while I get the message

`QProcess: Destroyed while process ("/usr/bin/kpackagetool5") is still running.`

In the widget panel the installation seems to work. I can see the application, but when I drag and drop it to the desktop I get the following message

![Screenshot_20240311_215104](https://github.com/blackadderkate/weather-widget-2/assets/7189867/4c33252a-458c-439a-98e8-9a172108ef0f)

Regardless of whether the installation is done via the command line or the widget panel, the folder under `~/.local/share/plasma/plasmoids` looks incomplete

```
└── plasmoids
    └── org.kde.weatherWidget-3
        ├── contents
        │   └── code
        │       ├── config-utils.js
        │       ├── data-loader.js
        │       ├── db
        │       │   ├── AD.csv
        │       │   ├── AE.csv
        │       │   ├── AF.csv
        <--- more... --->
        │       │   ├── UG.csv
        │       │   ├── US.csv
        │       │   └── UY.csv
        │       ├── icons.js
        │       ├── model-utils.js
        │       ├── placesearch-helpers.js
        │       ├── timezoneData.js
        │       └── unit-utils.js
        ├── metadata.json
        └── weather-widget.svg


```
If I create my own installation package (even though this is not recommended), it works ;) Files are then saved in a system-wide folder and not in the local user folder. And... once I create an instance of the widget on the desktop, remove it and uninstall the package, then the installation via the widget panel also works; then after a few minutes the files will be copied to the local user folder.

I also tried with another widget from the store to check if it is possible to install from the widget panel and there is no error in the current Plasma version and it works. I'm unsure if I've done something wrong or if your plasmoid archive is having problems, I'm just reporting my observations. You may have the same experience if you reinstall it in a new/clean environment.

For now it's fine for me to use my install package.

**System-Information:**
_Operating System: Arch Linux 
KDE Plasma Version: 6.0.2
KDE Frameworks Version: 6.0.0
Qt Version: 6.6.2
Kernel Version: 6.7.9-arch1-1 (64-bit)
Graphics Platform: X11_

**Reply by blackadderkate** [reply]
@martin-ms 
I've had a look at the plasmoid Beta Release and all the files are definitely present.

I _suspect_ that plasmapkg2 is incompatible with Plasma 6 widgets, causing the installer to crash halfway through - probably because metadata.desktop is not present in the plasmoid file as it is obsolete in Plasma6.

If you didn't remove this failed installation before using "Add Widget", it _might_ think it is already installed - and because the Version Number being installed matches the one in the existing Metadata.json file it might think it doesn't need upgrading.

Just a theory!

Maybe you can try fully removing the ~/.local/share/plasma/plasmoids/org.kde.weatherWidget-3 folder, and Logging Off, Rebooting, or just restarting Plasmashell (using `plasmashell --replace &`) and then installing using "Add Widget"...

I'll have a hunt around and see if there's any documentation on how to install plasmoids using the command line in Plasma 6.

**Reply by martin-ms** [reply]
I'm afraid I was too impatient. In my test environment (VM, HDD, active snapshots) it takes about 45 seconds until all (~320) files are copied to the target folder. So if you try to call it immediately after installation, it fails because there are still missing files. If you try it after a minute, it works.

This will also be the reason why the installation times out on the command line because the installation was not completed in the allotted time and the process then ended before all files were copied.

Conclusion: the installation via the widget panel works if you wait for the files to be copied after the installation. Installing via the command line doesn't work because you run into a timeout and then an incompletely filled folder is left behind. Since you cannot change the timeout, I think installing via the command line is unsuitable on slow systems.

**Comment by SpencerKersey** [comment]
I think I found a small bug for the hover function in the widget. When hovering over the widget without it being open, it shows the currently selected city as seen below:
![image](https://github.com/blackadderkate/weather-widget-2/assets/77792710/0c18b5cc-2dd0-4b17-8634-91f1889e8b0d)
But, when you have the widget open and then hover over the widget icon, it will show Leicester even though I removed that default city from the list in the settings.
![image](https://github.com/blackadderkate/weather-widget-2/assets/77792710/f3387c4c-cf50-4fee-9b63-09b68c648a10)

Other than this tiny issue, everything seems to be running great from my limited use. Keep up the great work :)

**Reply by blackadderkate** [reply]
OK, thanks for letting me know. To be fair, the "Add/Remove/Reorder places" code _is_ on my ToDo List. 
Oh, if you have a step-by-step guide to reproducing the issue it'd be incredibly useful...

**Comment by Whyhow314** [comment]
I've been testing out scaling/hidpi behavior a bit. It looks like on both X11 and wayland the scaling is currently broken, but can be fixed by removing the `Screen.devicePixelRatio` multipliers from `FullRepresentation.qml`, `Meteogram.qml`, and `NextDayItem.qml`.

**Comment by blackadderkate** [comment]
> I've been testing out scaling/hidpi behavior a bit. It looks like on both X11 and wayland the scaling is currently broken, but can be fixed by removing the `Screen.devicePixelRatio` multipliers from `FullRepresentation.qml`, `Meteogram.qml`, and `NextDayItem.qml`.

Cool, thanks for letting me know.
I'll remove those multipliers for the next beta - hopefully this weekend.

**Comment by blackadderkate** [comment]
This week I have been working on getting the widget working when compiled, and installed for all users - and as a System Tray applet.

I spent literally two days trying to figure out why clicking the widget didn't bring up the "Full Representation" when in the System Tray, but it did as an ordinary widget in the panel... 

Todo:

1. Upload my "FullRepresentation" fixes for the system tray.
2. Implement the fixes for Scaling / High DPI as suggested by @Whyhow314 .
3. Implement fixes for when the user adds/deletes a location or changes the location order (which I think is the cause of @SpencerKersey 's bug).
4. See if I can fix screen corruption when you change the order in the Places list.

Hoping to release another beta this weekend - assuming items 3 and 4 don't give me any grief!

**Comment by begonia208** [comment]
I"m using the Weather Widget 3 (3.0.4) for quite some time now.
Imho it is by far the most beautiful and useful weather widget out there.
I have not experiences any bugs.
Therefor in my opinion it is high time that this masterpiece is published in the kde store!
